File: packages/extension-host/src/extension-host/devtools.ts
import { invoke } from "@tauri-apps/api/core";

// Dev-shell window actions. Silo suppresses the webview's native context menu
// app-wide (see menu-controller.ts), so the two things that menu used to offer —
// Reload and Inspect Element — move to the Window menu in dev builds
// (core.menu, gated on `import.meta.env.DEV`). Host-owned: they touch the
// platform (`invoke`) and the global window, so they're not capabilities for
// silo.*/third-party extensions — hence the `@silo-code/extension-host/internal` barrel.

/** Reload the webview — the native menu's "Reload". */
export function reloadWindow(): void {
  window.location.reload();
}

/**
 * Open the webview devtools — the native menu's "Inspect Element". Backed by
 * the `open_devtools` Tauri command, which is a no-op outside debug/`devtools`
 * builds, so this is safe to call unconditionally.
 */
export function openDevtools(): void {
  void invoke("open_devtools");
}

File: apps/desktop/src-tauri/src/commands/devtools.rs
//! Dev-shell devtools control. The frontend suppresses the webview's native
//! context menu app-wide (so Silo owns every menu), which also removes the
//! native "Inspect Element". In debug builds that action moves to the Window
//! menu and routes through this command — the JS API has no runtime devtools
//! control, only a build-time config flag.

/// Open the webview devtools ("Inspect Element"). Available in debug builds or
/// when the `devtools` feature is enabled; a no-op otherwise, so the frontend
/// can call it without guarding on the build profile.
#[tauri::command]
pub fn open_devtools(window: tauri::WebviewWindow) {
    #[cfg(any(debug_assertions, feature = "devtools"))]
    window.open_devtools();
    #[cfg(not(any(debug_assertions, feature = "devtools")))]
    let _ = window;
}
